### PR TITLE
[#5][Chore] Setup scss and style framework

### LINKED ISSUE #5: [Chore] Setup scss and style framework
## Why

In order to save time on UI tasks and keep the code clean.

## Insight 

- Configure the project to use SCSS from [this source](https://create-react-app.dev/docs/adding-a-sass-stylesheet).
- Choose to include bootstrap of tailwind library.
- Setup default variables in order to get closer to the targeted UI
 
## Who Benefits?

Developers.

### PR DESCRIPTION
**Pull request by malparty**
- #5 

## What happened 👀

- [x] Setup Bootstrap and use minimal SCSS components
- [x] Build default app layout to validate the good integration of Bootstrap

Extra:
- Add `yarn test:coverage` command to know the unit tests coverage
- Upgrade all packages with `yarn upgrade-interactive` to practice from the training

## Insight 📝

Describe in detail how to test the changes, which solution you tried but did not go with, referenced documentation is welcome as well.

> Note that the components tests will be built in #6 and #7 respectively. So that we can validate them via the CI/CD directly.

## Proof Of Work 📹

|Bootstrap variable rewrite is working | Default layout utilizes bootstrap colors|
|---|---|
|image|image|

Files sizes remain good on build — which confirms we do not import the whole bootstrap lib.

image

### CODE REVIEW

File: src/assets/stylesheets/_variables.scss [hunk 1]
```diff
@@ -0,0 +1,11 @@
+$black: #fff !default;
```
Comment (rafayet-monon) [inline src/assets/stylesheets/_variables.scss]: This doesn't look right? 🤔 Should this be white?

File: src/assets/stylesheets/_variables.scss [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+$gray-900: #f8f9fa !default;
+$gray-800: #e9ecef !default;
+$gray-700: #dee2e6 !default;
+$gray-600: #ced4da !default;
+$gray-500: #adb5bd !default;
+$gray-400: #6c757d !default;
+$gray-300: #495057 !default;
+$gray-200: #343a40 !default;
+$gray-100: #212529 !default;
+$white: #000 !default;
```
Comment (rafayet-monon) [inline src/assets/stylesheets/_variables.scss]: Maybe black?

File: src/assets/stylesheets/vendor/bootstrap/index.scss [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// Configuration
+@import 'bootstrap/scss/functions';
+@import 'bootstrap/scss/variables';
+@import 'bootstrap/scss/mixins';
+@import 'bootstrap/scss/utilities';
+
+// Re-enable these if needed.
+// Do not forget to check the order of import components before re-enable.
+// Layout & components
+@import 'bootstrap/scss/root';
```
Comment (rafayet-monon) [inline src/assets/stylesheets/vendor/bootstrap/index.scss:8]: How about keeping the uncommented imports together rather than making them scattered?

File: src/components/BackgroundImage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import React, { PureComponent } from 'react';
+import defaultImage from '../../assets/images/default-background.jpg';
+
+export default class BackgroundImage extends PureComponent {
```
Comment (rafayet-monon) [inline src/components/BackgroundImage/index.tsx:4]: Just curious, what is the reason behind using PureComponent?

File: src/assets/stylesheets/_variables.scss (hunk 1 above)
Comment (malparty) [reply src/assets/stylesheets/_variables.scss]: Wow, that was a bad approach to dark mode~ 
Fixed in 755d118 ;-)

File: src/assets/stylesheets/_variables.scss (hunk 2 above)
Comment (malparty) [reply src/assets/stylesheets/_variables.scss]: Same here, fixed in 755d118 ✅

File: src/assets/stylesheets/vendor/bootstrap/index.scss (hunk 3 above)
Comment (malparty) [reply src/assets/stylesheets/vendor/bootstrap/index.scss:8]: ✅ Good idea, updated in 6e5e697 :)

File: src/components/BackgroundImage/index.tsx (hunk 4 above)
Comment (malparty) [reply src/components/BackgroundImage/index.tsx:4]: I'm not 100% sure as I want to explore more this topic when I will have a routing setup.
From my readings, supposing later I have a prop providing another image, I need this component to re-render when the image change. Using PureComponent would compare the primitive value (ex: Image URL) instead of the object reference, so that if I have 2x different pages with a prop object different, it would NOT re-render that component, thus being better for performances.

It was explained in the [advanced react training](https://app.pluralsight.com/library/courses/reactjs-advanced/table-of-contents) on Pluralsight + deeper explanations can be found [here](https://codeburst.io/when-to-use-component-or-purecomponent-a60cfad01a81). 

But it's maybe too early as for now there is no state/prop usage ^^. Let me know if you have a recommendation for this case :) Thanks,

File: package.json [hunk 5]
```diff
@@ -20,12 +20,14 @@
     "@types/react-dom": "^17.0.0",
     "@typescript-eslint/eslint-plugin": "^4.5.0",
     "@typescript-eslint/parser": "^4.5.0",
+    "axios": "^0.23.0",
```
Comment (carryall) [inline package.json]: Can we use a specific version for this dependency?
```suggestion
    "axios": "0.23.0",
```

File: package.json [hunk 6]
```diff
@@ -20,12 +20,14 @@
...
     "babel-eslint": "^10.1.0",
     "babel-jest": "^26.6.0",
     "babel-loader": "^8.1.0",
     "babel-plugin-named-asset-import": "^0.3.7",
     "babel-preset-react-app": "^10.0.0",
     "bfj": "^7.0.2",
+    "bootstrap": "^5.1.3",
```
Comment (carryall) [inline package.json]:
```suggestion
    "bootstrap": "5.1.3",
```

File: src/components/BackgroundImage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import React, { PureComponent } from 'react';
+import defaultImage from '../../assets/images/default-background.jpg';
```
Comment (carryall) [inline src/components/BackgroundImage/index.tsx]: This might be a bit off-topic but how about adding `resolve.alias` on [webpack config](https://webpack.js.org/configuration/resolve/#resolvealias) so instead of relative path import you can use the alias module. Note that this can be in another PR 😄 
```
resolve: {
    alias: {
      @assets: path.resolve(__dirname, 'src/assets/'),
    }
  }
```

```js
import defaultImage from '@assets/images/default-background.jpg';
```

Comment (malparty) [reply src/components/BackgroundImage/index.tsx]: OMG I'm in love with this!!! In the past I had the chance to contribute on a vue project and I did love the fact that src folder was mapped to `@/...`, which I did not know how to do with React... And you just gave me the solution !! 🤪
I'll do it in this PR as it still makes sense while not such a big change (but huge value for dev experience :) )
Thanks!!!

File: package.json (hunk 5 above)
Comment (malparty) [reply package.json]: Resolved in the rebase :)

File: package.json (hunk 6 above)
Comment (malparty) [reply package.json]: Resolved in the rebase :)

File: src/components/BackgroundImage/index.tsx (hunk 7 above)
Comment (malparty) [reply src/components/BackgroundImage/index.tsx]: ✅ Implemented in 041a215
There is a small trap with Jest as it requires its own configuration for aliases~

File: src/assets/stylesheets/application.scss [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+// Dependencies
+@import 'vendor/bootstrap/index';
+
+// Functions
+@import 'functions/sizing';
+
+// Mixins
+
+// Base
```
Comment (carryall) [inline src/assets/stylesheets/application.scss:12]: How about adding the base font?

```suggestion
// Base
@import 'base/fonts';
@import 'base/layouts';
```

```scss
// base/fonts.scss
@font-face {
  font-family: 'Neuzeit S LT Std';
  ...
}

// base/layouts.scss
body {
  font-family: 'Neuzeit S LT Std', serif;
}
```

**Review by carryall — COMMENTED** [review]
One last comment, the rest LGTM 🙌🏼

File: src/assets/stylesheets/application.scss (hunk 8 above)
Comment (malparty) [reply src/assets/stylesheets/application.scss:12]: Good idea :) 
Done in 5f5521d ✅

File: src/assets/stylesheets/_variables.scss [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+$gray-100: #f8f9fa !default;
+$gray-200: #e9ecef !default;
+$gray-300: #dee2e6 !default;
+$gray-400: #ced4da !default;
+$gray-500: #adb5bd !default;
+$gray-600: #6c757d !default;
+$gray-700: #495057 !default;
+$gray-800: #343a40 !default;
+$gray-900: #212529 !default;
+$black: #000;
```
Comment (carryall) [inline src/assets/stylesheets/_variables.scss:11]: Sorry, @malparty I just notice that this color is not as black as its name 🙈
Screen Shot 2564-10-19 at 18 35 50

Comment (malparty) [reply src/assets/stylesheets/_variables.scss:11]: As the application is still using true-black for the background gradient, I will keep $black=#000; but I did set that shade of dark gray as $gray-900.

I took the opportunity to fix the background image style in order to have a gradient like from the mockup :)

✅ Done in 08985c2 :)

**Review by carryall — APPROVED** [review]

File: src/assets/stylesheets/_variables.scss [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+$gray-600: #6c757d !default;
+$gray-700: #495057 !default;
+$gray-800: #343a40 !default;
+$gray-900: #15151a !default;
+$black: #000;
+
+$body-bg: rgba(0, 0, 0, 0);
+$body-color: $white;
+
+$zindex-background-image: -1;
```
Comment (rafayet-monon) [inline src/assets/stylesheets/_variables.scss]: how about 
```suggestion
$zindex-behind: -1;
```

**Review by rafayet-monon — APPROVED** [review]
Jus one comment

File: src/assets/stylesheets/_variables.scss (hunk 10 above)
Comment (malparty) [reply src/assets/stylesheets/_variables.scss]: ✅ 7ce1d91
Good idea, I like it!! Thanks :)
